File: src/core/platform.ts
import { existsSync, readFileSync } from "node:fs";
import { HubInstallerError } from "../errors";
import type { SupportedPlatform } from "../types";

function readLinuxOsRelease(): string {
  const releasePath = "/etc/os-release";
  if (!existsSync(releasePath)) {
    return "";
  }

  try {
    return readFileSync(releasePath, "utf8").toLowerCase();
  } catch {
    return "";
  }
}

function detectLinuxPlatform(): SupportedPlatform {
  const osRelease = readLinuxOsRelease();

  if (osRelease.includes("ubuntu") || osRelease.includes("debian")) {
    return "ubuntu";
  }

  return "ubuntu";
}

export function detectHostPlatform(): SupportedPlatform {
  switch (process.platform) {
    case "win32":
      return "windows";
    case "darwin":
      return "macos";
    case "linux":
      return detectLinuxPlatform();
    default:
      throw new HubInstallerError(
        "UNSUPPORTED_HOST",
        `Unsupported host platform: ${process.platform}`
      );
  }
}

File: rust/src/platform.rs
use std::{env, fs};

use crate::{
    error::{HubError, Result},
    types::SupportedPlatform,
};

pub fn detect_host_platform() -> Result<SupportedPlatform> {
    match env::consts::OS {
        "windows" => Ok(SupportedPlatform::Windows),
        "macos" => Ok(SupportedPlatform::Macos),
        "linux" => Ok(detect_linux_platform()),
        other => Err(HubError::message(
            "UNSUPPORTED_HOST",
            format!("unsupported host platform: {other}"),
        )),
    }
}

fn detect_linux_platform() -> SupportedPlatform {
    let _os_release = fs::read_to_string("/etc/os-release")
        .unwrap_or_default()
        .to_lowercase();
    SupportedPlatform::Ubuntu
}
